redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/handler/AllocateDatabaseServerHandler.java

```diff
@@ -48,40 +61,110 @@
         try {
             CloudConnector<Object> connector = cloudPlatformConnectors.get(cloudContext.getPlatformVariant());
-            AuthenticatedContext ac = connector.authentication().authenticate(cloudContext, request.getCloudCredential());
-            List<CloudResourceStatus> resourceStatuses =
-                connector.resources().launchDatabaseServer(ac, request.getDatabaseStack(), persistenceNotifier);
+            CloudCredential cloudCredential = request.getCloudCredential();
+            AuthenticatedContext ac = connector.authentication().authenticate(cloudContext, cloudCredential);
+            DatabaseStack databaseStack = request.getDatabaseStack();
+            prescribeSslCertificateIfNeeded(connector, cloudContext, cloudCredential, request.getDbStack(), databaseStack.getDatabaseServer());
+            List<CloudResourceStatus> resourceStatuses = connector.resources().launchDatabaseServer(ac, databaseStack, persistenceNotifier);
             List<CloudResource> resources = ResourceLists.transform(resourceStatuses);
 
             PollTask<ResourcesStatePollerResult> task = statusCheckFactory.newPollResourcesStateTask(ac, resources, true);
             ResourcesStatePollerResult statePollerResult = ResourcesStatePollerResults.build(cloudContext, resourceStatuses);
             if (!task.completed(statePollerResult)) {
-                syncPollingScheduler.schedule(task);
+                statePollerResult = syncPollingScheduler.schedule(task);
             }
-            RedbeamsEvent success = new AllocateDatabaseServerSuccess(request.getResourceId());
+            validateResourcesState(cloudContext, statePollerResult);
+            response = new AllocateDatabaseServerSuccess(request.getResourceId());
             // request.getResult().onNext(success);
-            eventBus.notify(success.selector(), new Event<>(event.getHeaders(), success));
             LOGGER.debug("Launching the database stack successfully finished for {}", cloudContext);
         } catch (Exception e) {
-            AllocateDatabaseServerFailed failure = new AllocateDatabaseServerFailed(request.getResourceId(), e);
+            response = new AllocateDatabaseServerFailed(request.getResourceId(), e);
             LOGGER.warn("Error launching the database stack:", e);
             // request.getResult().onNext(failure);
-            eventBus.notify(failure.selector(), new Event<>(event.getHeaders(), failure));
         }
+        return response;
     }
+
+    @Override
+    protected Selectable defaultFailureEvent(Long resourceId, Exception e, Event<AllocateDatabaseServerRequest> event) {
+        AllocateDatabaseServerFailed failure = new AllocateDatabaseServerFailed(resourceId, e);
+        LOGGER.warn("Error launching the database stack:", e);
+        return failure;
+    }
+
+    private void prescribeSslCertificateIfNeeded(CloudConnector<Object> connector, CloudContext cloudContext, CloudCredential cloudCredential, DBStack dbStack,
```

Comment:
this method is pretty complex. probably some refactor does not really hurt

Reply:
Good point, has been broken up into multiple methods.